Shift configured parameters to a new Asciidoc

vihaanthora opened the pull request:
This pull request will complete the below tasks - 
- [x] Create a processing layer for separating the configured parameters from Pipeline steps documentation.
- [x] Update `README.md` with instructions that are specific to adding parameters to the config file.
- [x] Add tests for `ProcessAsciiDoc.java`.
- [x] Adding comments to the code if required.

I will add some notes to [this](https://issues.jenkins.io/browse/WEBSITE-809) ticket explaining some logical decisions taken while writing this code. Meanwhile, feel free to review the code and suggest any possible improvements.

cc: @kwhetstone

kwhetstone commented on README.md:
```diff
@@ -113,3 +113,13 @@ make -C ../jenkins.io run
+
+### 6. Configure Parameters
+
+The parameters that need to be separated to new pages can be entered in `config.txt` by adhering to the following rules.
```
You could also add another sentence about why you're making this change.  Just saying "this prevents having really long page loads" would be really helpful.

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+        File[] directoryListing = dir.listFiles();
+        try {
+            List<String> config = Files.readAllLines(Paths.get("config.txt"));
+            for (String param : config) {
```
I might use a different variable name other than `param` like `className`, just something a bit clearer about what is coming from the file.

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+        return duplicate.toString();
+    }
+
+    public void processDocs(String allAscii) {
```
Add a javadoc comment about what is being done generally by this class & this function.  Note that failing this method will return out.  Does this throw an error?  You should note that with a `throws` in Java standard format.

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+        BufferedReader br = new BufferedReader(new FileReader(child));
+        String line;
+        int counter = 0; // keeps a count of opening and closing li tags to mark the end of a section
+        boolean flag = false;
```
Is this just used to mark in the bufferedlinereader if this is the start of a section?

vihaanthora replied:
We are marking the end of a section when the `counter` becomes zero again and `flag` is true.

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+     */
+    public String separateParam(String param, String allAscii, File child)
+            throws IOException, RuntimeException {
+        BufferedReader br = new BufferedReader(new FileReader(child));
```
I know you've declared that this throws an `IOException`, but you should catch this here since you're working with the file, and you can close the FileReaders in a finally statement.

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /*
+     * Takes in param name, and separates its documentation on to a new file.
```
Let's detail what a "param" is here.  Everything that goes into a function is a param :)

kwhetstone commented on src/main/java/org/jenkinsci/pipeline_steps_doc_generator/ProcessAsciiDoc.java:
```diff
@@ -0,0 +1,93 @@
+                duplicate.append(line).append("\n");
+            }
+            if (flag) {
+                counter += (line.split("<li>", -1).length - line.split("</li>", -1).length);
```
I'm not 100% sure what this is doing?  Is it looking for open and close tags?

vihaanthora replied:
Yes, it is adding the difference between the number of opening and closing tags to see when a logical sequence of tags ends.

kwhetstone reviewed (commented):
Added some comments about the code.  I think you could rename some of the variables to make it easier to see what's going on.

I think for the tests, you've got the right start.  It would be good to also check if the link replacement in the original file is done correctly.  I think you should be able to move this to a full PR after addressing these comments.

vihaanthora commented:
Thanks for the review @kwhetstone. I will address these and push a commit. We can also discuss any other changes in the office hours today.

vihaanthora commented:
@kwhetstone I've added the tests that you suggested in our last meeting, anything else remaining to complete this PR?

vihaanthora commented:
While looking for class names that should be included in the config file in the beginning, I noticed that very few have documentation more than 500 lines long, so it would be better to reduce the threshold to maybe 100.

vihaanthora commented:
I have added a configuration that I found suitable to begin with, and the results are very encouraging. `workflow-multibranch.adoc` is only 688 kBs now! I've also added some comments in the config file about the classes included.

kwhetstone reviewed (approved):
Great!  This looks good, and I appreciate the tests.  Let's merge and look at the new screenshots.

MarkEWaite commented:
This pull request seems to be the cause of broken site generation on the jenkins.io site.  The [master branch](https://ci.jenkins.io/job/Websites/job/jenkins.io/job/master/) fails to build as do other branches like the [2.367 changelog](https://ci.jenkins.io/job/Websites/job/jenkins.io/view/change-requests/job/PR-5396/), a [documentation improvement](https://github.com/jenkins-infra/jenkins.io/pull/5390), and the [2.366 changelog](https://github.com/jenkins-infra/jenkins.io/pull/5379).

I'll prepare a revert pull request for discussion in case the issue needs more time for @vihaanthora to investigate the cause.  We release Jenkins 2.366 today but the changelog is not available because the site cannot be generated.

vihaanthora commented:
Thanks for sharing the issue @MarkEWaite. I have started working on it. Will create a PR as soon as it gets resolved.

vihaanthora commented:
@MarkEWaite I have found the cause of this error; it was that I did not write code to create the `params` directory, and had created it manually on my local machine when I was working on developing the processing layer. Also, I did not check the contents of allAscii.zip assuming that the result will be the same as the local copy. Due to this, the processing layer was having to exit whenever it was told to write in the `params` folder, resulting in incomplete asciidocs being written. 

I'll add that line and create a PR. I'll then take the generated allAscii.zip and test it locally before merging. 

Apologies for not checking allAscii.zip before asking to merge this PR.

MarkEWaite commented:
Thanks so much for your contributions @vihaanthora !  There will always be bumps and bruises while improving things.  Your work is such an improvement to the project!
